documentation/writing_tests.md

```diff
@@ -16,4 +16,29 @@
 
 ## Structure of tests
 
-## Individual tests
\ No newline at end of file
+## Writing test cases
+
+a. On titles
+	i. Test should only test one thing. The titles should reflect that.
+b. Each test should just test one function or method.
+	i. Constructors are a function.
+	ii. Tests can build on top of previous tests.
+		1) One test can validate a constructor while another can use the constructor, then validate a method of the instantiated object.
+	iii. The structure then should be:
+		1) Define what is needed for this method to run.
+		2) Run the method / function clearly, just once.
+		3) Test the results of that method.
+c. On the usage of conditionals
+	i. Tests can be skipped if necessary.
+	ii. If a test runs, it should always work the same way, not behave differently depending on the environment.
+	iii. If you still need to do conditionals in tests, copy these tests as many times as needed to remove the conditionals.
+d. Unit tests should not use the recorder.
+	i. They should not reach out to real services.
+	ii. Unit tests might use mocking tools that could interfere with the recorder. If that's the case, avoid running these unit tests if TEST_MODE is set.
+	iii. Unit tests should not alter a global or shared state.
+e. Recorded and live tests should look like samples.
+	i. Group the live tests on what they're testing.
+	ii. Test all the possible exceptions. Group the exceptions in the same bucket of the separation of tests by what they're testing, but at the end of that/those files.
```

Comment:
`all possible exceptions` seems a bit large, technically the JS runtime can throw an out of memory exception, but I don't think we're talking about that. Also I'm sure there are certain service exceptions that are difficult to force. Perhaps this means exercising client-side validation inside methods? e.g. validating batch size in storage.

Reply:
Oh! I forgot to add a revision for this. Apparently this can be done through TypeDocs instead of tests. Using TypeDocs to document exceptions seems far better than having to test exceptions, at least for me.

Reply:
I have updated this PR with more concrete information. I will come back tomorrow to go through  all of my notes and see if I'm missing anything.